src/ast_parser.rs

```diff
@@ -133,83 +102,60 @@ pub(crate) struct AstParser {
 
 impl AstParser {
   pub(crate) fn new() -> Self {
-    let buffered_error = SwcErrorBuffer::default();
-
-    let handler = Handler::with_emitter_and_flags(
-      Box::new(buffered_error.clone()),
-      HandlerFlags {
-        dont_buffer_diagnostics: true,
-        can_emit_warnings: true,
-        ..Default::default()
-      },
-    );
-
     let globals = Globals::new();
     let top_level_mark =
       swc_common::GLOBALS.set(&globals, || Mark::fresh(Mark::root()));
 
     AstParser {
-      buffered_error,
-      source_map: Rc::new(SourceMap::default()),
-      handler,
       globals,
       top_level_mark,
     }
   }
 
-  pub(crate) fn set_source_map(&mut self, source_map: Rc<SourceMap>) {
-    self.source_map = source_map;
-  }
-
   pub(crate) fn parse_program(
     &self,
     file_name: &str,
     syntax: Syntax,
     source_code: &str,
-  ) -> Result<ParsedData, SwcDiagnosticBuffer> {
-    // NOTE: calling `self.source_map.new_source_file` mutates `source_map`
-    // even though it's of type `Rc`.
-    let swc_source_file = self.source_map.new_source_file(
-      FileName::Custom(file_name.to_string()),
-      source_code.to_string(),
+  ) -> Result<ParsedData, SwcDiagnostic> {
+    let source_file =
+      SourceFileTextInfo::new(BytePos(0), source_code.to_string());
+    let string_input = StringInput::new(
+      source_code,
+      BytePos(0),
+      BytePos(source_code.len() as u32),
     );
 
-    let buffered_err = self.buffered_error.clone();
-
     let comments = SingleThreadedComments::default();
-    let lexer = Lexer::new(
+    let lexer = Capturing::new(Lexer::new(
       syntax,
       JscTarget::Es2019,
-      StringInput::from(&*swc_source_file),
+      string_input,
       Some(&comments),
-    );
-
-    let tokens: Vec<TokenAndSpan> = lexer.clone().into_iter().collect();
+    ));
 
     let mut parser = Parser::new_from(lexer);
+    let program = parser.parse_program().map_err(|err| {
+      SwcDiagnostic::from_swc_error(file_name, &source_file, err)
+    })?;
 
-    let parse_result = parser.parse_program().map_err(move |err| {
-      let mut diagnostic_builder = err.into_diagnostic(&self.handler);
-      diagnostic_builder.emit();
-      SwcDiagnosticBuffer::from_swc_error(buffered_err, self)
+    let program = swc_common::GLOBALS.set(&self.globals, || {
+      program.fold_with(&mut ts_resolver(self.top_level_mark))
     });
```

Comment:
What does this do?

Reply:
This is used to apply proper "syntax context" to all AST elements - when SWC performs transforms/folding it might change some of those context and "ts_resolver" ensures that all elements end up in proper lexical scope.